Add a script to calculate scores from results

tbalacha opened the pull request:
This script uses results from .gb3 files to calculate Geekbench scores and rates.

tbalacha committed:
Add a script to calculate scores from results

aschmied commented on geekbench-calculate-scores.py:
```diff
@@ -0,0 +1,148 @@
		return int(2500 * rel)
	
+	def wl_mc_score(self):
+		rel = self.mc_rate() / baseline_rate[self.name]
```
You can get rid of some of the chained indexing operations by doing something like:

```
self.sc_runtime = 0
self.sc_work = 0
self.mc_runtime = 0
self.mc_work = 0

for result in jsonf['results']:
  if result['threads'] == 1:
    self.sc_runtime = result['runtime']
    self.sc_work = result['work']
  elif result['threads'] > 1:
    self.mc_runtime = result['runtime']
    self.mc_work = result['work']
  else:
    raise ValueError("Result had zero worker threads for {}".format(name))
```

aschmied commented on geekbench-calculate-scores.py:
```diff
@@ -0,0 +1,148 @@
+	divisor = 1000.0

+	if (wl_unit == "B/sec"):
+		divisor = 1024.0
```
This is the only place you use scipy, so I recommend computing the geomean yourself to remove the dependence on scipy. It will make the script more portable since scipy is a non-standard package.

tbalacha committed:
Fix line endings

tbalacha committed:
Remove extra spaces

tbalacha committed:
Replace all tabs with spaces

tbalacha committed:
Modify repetitive indexing operations

tbalacha committed:
Add geometric mean calculation to remove scipy dependency

tbalacha committed:
Remove debugging output

aschmied commented on geekbench-calculate-scores.py:
```diff
@@ -0,0 +1,221 @@
+
+  print "\t%s%s%s".expandtabs(2) % ("Geekbench Score".ljust(24),
+  gb_sc_score_just, gb_mc_score_just)
+  jsonfile.close()
```
It's a good idea to close the file as soon as you're done reading it. You could do it immediately after the `jsonobj = json.load(jsonfile)` line.

aschmied commented on geekbench-calculate-scores.py:
```diff
@@ -0,0 +1,221 @@
+    for wl in wls:
+      print "\t%s".expandtabs(2) % wl.name
+
+      wl_sc_score_str = str(int(wl.wl_sc_score())).rjust(12)
```
`wl_sc_score` already returns an `int` so you don't need the `int` conversion here

aschmied commented on geekbench-calculate-scores.py:
```diff
@@ -0,0 +1,221 @@
+      sc_rate_str = get_rate_string(wl.name, wl.sc_rate()).rjust(21)
+      print "\tsingle-core%s%s".expandtabs(4) % (wl_sc_score_str, sc_rate_str)
+
+      wl_mc_score_str = str(int(wl.wl_mc_score())).rjust(13)
```
`wl_sc_score` already returns an `int` so you don't need the `int` conversion here

aschmied commented on geekbench-calculate-scores.py:
```diff
@@ -0,0 +1,221 @@
+  for score in scores:
+    score_total *= score
+  nth_root = 1 / float(len(scores))
+  return int(math.pow(score_total, nth_root))
```
The `int` conversion should probably go in `compute_sect_sc_score` and `compute_sect_mc_score` since conversion to `int` is part of computing the score, not the geometric mean. As it's written this function is technically not computing the geometric mean; it is computing the geometric mean with the fractional part truncated.

tbalacha committed:
Fix integer conversions

This commit moves the int conversion from geomean() to the
compute_sect_*_score functions for accuracy.
There were also unnecessary int conversions in main().

aschmied commented on geekbench-calculate-scores.py:
```diff
@@ -0,0 +1,222 @@
+    print ""
+
+  print "Benchmark Summary"
+  for section in sections.iterkeys():
```
It's not critical for this PR, but you can iterate over `dict` keys and values together using `iteritems()`. Something like

```
for (section_name, section) in sections.iteritems():
  section_score_str = str(section_name + " Score").ljust(24)
  sect_sc_score_str = str(compute_sect_sc_score(section)).rjust(7)
```

Just something to note for the future.
